File: Hito_4/Observer/VehicleRoutingOriginal/VehicleRoutingSolucionadoObserver/VehicleRoutingValidator/Handlers/DepotIsNotVisitedTwiceInARowHandler.cs
using VehicleRoutingValidator.VehicleRouting;

namespace VehicleRoutingValidator.Handlers;

public class DepotIsNotVisitedTwiceInARowHandler : BaseHandler
{
    private const string Description = "Depot is visited twice in a row.";
    private const int Depot = VehicleRoutingProblem.Depot;

    public override bool IsSolutionValid(int[] solution)
    {
        if(IsDepotNotVisitedTwiceInARow(solution))
            return base.IsSolutionValid(solution);
        UpdateSubscribers(Description);
        return false;
    }

    private bool IsDepotNotVisitedTwiceInARow(int[] solution)
    {
        for(int i = 1; i < solution.Length; i++)
            if (solution[i - 1] == Depot && solution[i] == Depot)
                return false;

        return true;
    }
}

File: Hito_4/Observer/VehicleRouting/handlers/DepotIsNotVisitedTwiceInARowHandler.py
import BaseHandler
from ..models.VehicleRoutingProblem import VehicleRoutingProblem

class DepotIsNotVisitedTwiceInARowHandler(BaseHandler):
    DESCRIPTION = "Depot is visited twice in a row."
    DEPOT = VehicleRoutingProblem.DEPOT

    def __init__(self):
        super().__init__()

    def is_solution_valid(self, solution: list[int]) -> bool:
        if self._is_depot_not_visited_twice_in_a_row(solution):
            return super().is_solution_valid(solution)
        self.update_subscribers(self.DESCRIPTION)
        return False

    def _is_depot_not_visited_twice_in_a_row(self, solution: list[int]) -> bool:
        for i in range(1, len(solution)):
            if solution[i - 1] == self.DEPOT and solution[i] == self.DEPOT:
                return False
        return True
